provenance/core.py

```diff
@@ -91,23 +96,29 @@ def fn_info(f):
     return info
 
 
-def hash_inputs(inputs):
+def hash_inputs(inputs, check_mutations=False):
     kargs = {}
     varargs = []
-    all_ids = frozenset()
+    all_artifacts = {}
 
     for k, v in inputs['kargs'].items():
-        h, ids = hash(v, hasher=ah.artifact_id_hasher())
+        h, artifacts = hash(v, hasher=ah.artifact_hasher())
         kargs[k] = h
-        all_ids |= ids
+        all_artifacts = t.merge(all_artifacts, artifacts)
 
     for v in inputs['varargs']:
-        h, ids = hash(v, hasher=ah.artifact_id_hasher())
+        h, artifacts = hash(v, hasher=ah.artifact_hasher())
         varargs.append(h)
-        all_ids |= ids
+        all_artifacts = t.merge(all_artifacts, artifacts)
+
+    if check_mutations:
+        for a in all_artifacts.values():
+            if a.value_id != hash(a.value):
+                msg = "Artifact {}, of type {} has been mutated"
```

Comment:
I would include the name of the parameter that has been mutated, we could do that with `for param_name, a in all_artifacts.items():`, right?  That way it would mirror the way the impure function error is reported.